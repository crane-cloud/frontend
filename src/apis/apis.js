import axios from "../axios";
import activityLoggerAxios from "./userActivityLoggerAxios";

export const handlePostRequestWithDataObject = (data, endpoint) => {
  return new Promise((resolve, reject) => {
    axios
      .post(endpoint, {
        data,
      })
      .then((response) => {
        resolve(response);
      })
      .catch((error) => {
        reject(error);
      });
  });
};
export const handlePostRequestWithOutDataObject = (data, endpoint) => {
  return new Promise((resolve, reject) => {
    axios
      .post(endpoint, data)
      .then((response) => {
        resolve(response);
      })
      .catch((error) => {
        reject(error);
      });
  });
};

export const handleGetRequest = (endpoint) => {
  return new Promise((resolve, reject) => {
    axios
      .get(endpoint)
      .then((response) => {
        resolve(response);
      })
      .catch((error) => {
        reject(error);
      });
  });
};

export const handleUserActivitiesGetRequest = (endpoint) => {
  return new Promise((resolve, reject) => {
    activityLoggerAxios
      .get(endpoint)
      .then((response) => {
        resolve(response);
      })
      .catch((error) => {
        reject(error);
      });
  });
};

export const handlePatchRequest = (endpoint, data) => {
  return new Promise((resolve, reject) => {
    axios
      .patch(endpoint, data)
      .then((response) => {
        resolve(response);
      })
      .catch((error) => {
        reject(error);
      });
  });
};
export const handleDeleteRequest = (endpoint, data) => {
  return new Promise((resolve, reject) => {
    axios
      .delete(endpoint, data)
      .then((response) => {
        resolve(response);
      })
      .catch((error) => {
        reject(error);
      });
  });
};
