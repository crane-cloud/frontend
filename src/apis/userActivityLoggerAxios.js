import axios from "axios";
import { ACTIVITY_LOGS_API_URL } from "../config";

const activityLoggerAxiosInstance = axios.create({
  baseURL: ACTIVITY_LOGS_API_URL,
});

activityLoggerAxiosInstance.defaults.headers.Authorization = `Bearer ${localStorage.getItem(
  "token"
)}`;

activityLoggerAxiosInstance.interceptors.response.use(
  (response) => response,
  (error) => {
    let e = error;
    // this checks for when a token is not verified and logs you out
    if (error.response.status === 401 || error.response.status === 422) {
      localStorage.clear();
      window.location.href = "/";
    } else if(error.response.status === 502){
      e = "Your request too long possible Server Error."
      window.location.href = "/projects";
    }
    return Promise.reject(e);
  }
);

export default activityLoggerAxiosInstance;
