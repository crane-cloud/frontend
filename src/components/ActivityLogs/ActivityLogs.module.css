.DetailsSection {
  display: flex;
  flex-direction: column;
  gap: 1rem;
  margin-bottom: 1rem;
}

.Header {
  display: flex;
  align-items: center;
  justify-content: space-between;
  flex-wrap: wrap;
  margin-bottom: 2rem;
}

.Heading {
  font-weight: bolder;
  line-height: 1.8;
  font-size: larger;
}

.DateSection {
  padding: 5px 0px 2px 0px;
  display: flex;
  flex-direction: row;
  font-size: 15px;
  justify-content: space-evenly;
  align-items: center;
  /* height: 25px; */
  /* gap: .5rem; */
  margin-top: 3px;
  position: relative;
}

.DateItem {
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
}

.FilterButton {
  outline: none;
  background: unset;
  border: 1px solid black;
  color: #000;
}

.FilterButton:hover {
  border: none;
}

.Filter {
  width: 20rem;
  padding: .3rem;
  border: 1px solid #000000;
  display: flex;
  padding-left: 7px;
  padding-right: 7px;
  z-index: 1;
  background-color: #fff;
  justify-content: space-between;
  align-items: center;
  border-radius: 9px;
}

.SimpleForm {
  display: flex;
  flex-direction: row;
  align-items: flex-end;
  position: relative;
  align-items: center;
  gap: .7rem;
}

.errorSection {
  font-size: 11px;
  color: red;
  display: block;
  align-self: center;
}

.FilterDropdown {
  width: 12.5rem;
  right: 0.65rem;
  height: 7rem;
  overflow-y: auto;
  position: absolute;
  top: 1.5rem;
  transition: all;
  z-index: 1;
  overflow-x: hidden;
  /* max-height: 23rem; */
  overflow: visible;
  border: 1px solid #000000;
  display: flex;
  background-color: #fff;
  flex-direction: column;
  align-items: center;
  border-radius: 9px;
}

.FilterText {
  font-weight: 400;
  font-size: 15px;
  text-align: center;
  color: #555555;
}

.FilterItemContainer {
  width: 12rem;
  font-size: 12px;
  font-weight: 500;
  border-bottom: 1px solid #ccc;
}

.FilterItem {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  padding: .4rem 1rem .8rem 1rem;
  position: relative;
}

.DropdownArrowSvg {
  width: 2rem;
  height: 2rem;
  display: flex;
  align-self: center;
  fill: black;
}

.OnHeader:hover .DropdownArrowSvg {
  fill: var(--secondary-color);
}

.RelativeContainer {
  position: relative;
}

.SelectorBox {
  width: 95%;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  height: 20px;
  padding: 1px 1px 0 3px;
  /* margin-top: 3px; */
  border-radius: 9px;
  border: 2px solid #000000;
}

.SelectOption {
  font-size: 13px;
}

.InnerDropDown {
  position: absolute;
  width: 7rem;
  max-height: 15rem;
  background-color: #e2f5ff;
  z-index: 11;
  overflow-y: auto;
  overflow-x: hidden;
  right: 30px;
  border: 1px solid #ccc;
  top: 46px;
  display: flex;
  box-shadow: 0 4px 4px 0 rgb(117 117 117 / 20%);
  flex-direction: column;
}

.InnerDropDownLowerhalf {
  position: absolute;
  max-width: 10rem;
  max-height: 15rem;
  background-color: #e2f5ff;
  z-index: 11;
  overflow-y: auto;
  overflow-x: hidden;
  right: 37px;
  border: 1px solid #ccc;
  bottom: 46px;
  display: flex;
  box-shadow: 0 4px 4px 0 rgb(117 117 117 / 20%);
  flex-direction: column;
}

.InnerDropDownItem {
  display: flex;
  text-align: start;
  padding: 10px 10px 2px 10px;
  border-bottom: 1px solid #ccc;
}

.InnerDropDownItem:hover {
  cursor: pointer;
  background-color: #c1e6fa;
}

.Table {
  background: #ffffff;
  border-radius: 15px;
  padding: 1rem 2rem;
}

.TableRow {
  display: flex;
  align-items: center;
  gap: 1.5rem;
}

.Success {
  color: green;
}

.Danger {
  color: red;
}

.hr {
  margin: .5rem 0;
}

.Row {
  display: flex;
  justify-content: space-between;
  width: -webkit-fill-available;
}

.Bold {
  font-weight: bold;
  font-size: 17px;
}

.RowCell {
  display: flex;
  gap: 1rem;
}

.LastCell {
  align-self: flex-end;
}

.CalenderFromposition {
  background-color: #ffffff;
  border: 1px solid #cccccc;
  /* padding: 0.5rem; */
  display: grid;
  /* gap: 1rem; */
  position: absolute;
  z-index: 10;
  left: 1rem;
  border-radius: 10px;
}

.CalenderToposition {
  background-color: #ffffff;
  border: 1px solid #cccccc;
  /* padding: 0.5rem; */
  display: grid;
  /* gap: 1rem; */
  position: absolute;
  z-index: 10;
  left: 6rem;
  border-radius: 10px;
}

.NoResourcesMessage {
  display: flex;
  justify-content: center;
  width: 100%;

}

.OuterFilterItem {
  background-color: #fff;
  border: 1px solid #000;
  border-radius: 9px;
  height: 40px;
  width: 100%;
  z-index: 1;
}