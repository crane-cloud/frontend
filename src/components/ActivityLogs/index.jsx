import styles from "./ActivityLogs.module.css";
import { handleGetRequest } from "../../apis/apis.js";
import { ReactComponent as CheckMark } from "../../assets/images/check-circle.svg";
import { ReactComponent as Danger } from "../../assets/images/alert-octagon.svg";
// import { ReactComponent as CloudOff } from "../../assets/images/cloud-off.svg";
// import { ReactComponent as Upload } from "../../assets/images/upload-cloud.svg";
import { ReactComponent as User } from "../../assets/images/user.svg";
import DateInput from "../../components/DateInput";
import React, { useState, useEffect, useRef } from "react";
import { ReactComponent as DownArrow } from "../../assets/images/downarrow.svg";
import { ReactComponent as FilterIcon } from "../../assets/images/filterIcon.svg";
import { ReactComponent as ArrowUpDDown } from "../../assets/images/ArrowUp&Down.svg";
import { DisplayDateTime } from "../../helpers/dateConstants";
import Spinner from "../Spinner";

const ActivityLogs = ({ projectID }) => {
  const [toTS, setToTS] = useState("none");
  const [fromTS, setFromTS] = useState("none");
  const [showStatusFilter, setShowStatusFilter] = useState(false);
  const [filterOpen, setFilterOpen] = useState(false);
  const [showToCalendar, setShowToCalendar] = useState(false);
  const [showFromCalendar, setShowFromCalendar] = useState(false);
  const [showOperation, setShowOperation] = useState(false);
  const [operationField, setOperationField] = useState("none");
  const [statusField, setStatusField] = useState("none");
  const [feedback, setFeedback] = useState("none");
  const [queryParams, setQueryParams] = useState("");
  const [dateError, setDateError] = useState("");

  const statusList = ["Success", "Failed"];
  const operationsList = ["Create", "Update", "Delete"];
  //api
  const [loading, setLoading] = useState(false);
  const [logs, setLogs] = useState([]);
  const baseLinkRef = useRef(`/users/activities?`);
  // const baseLink = baseLinkRef.current;
  useEffect(() => {
    if (projectID) {
      // logs based on a project
      // set filters available per page here
      baseLinkRef.current = `${baseLinkRef.current}a_project_id=${projectID}&`;
    }
  }, [projectID]);

  useEffect(() => {
    fetchActivityLogs(`${baseLinkRef.current}${queryParams}`);
    setFeedback("");
  }, [queryParams]);

  const fetchActivityLogs = (link) => {
    setLoading(true);
    handleGetRequest(link)
      .then((response) => {
        if (response.data.data.activity.length > 0) {
          setLogs(response.data.data.activity);
        } else {
          setFeedback("No logs for you");
        }
        setLoading(false);
      })
      .catch((error) => {
        setFeedback("Failed to fetch logs, please try again");
        setLoading(false);
      });
  };

  const handleStatusFilter = (item) => {
    setStatusField(item);
    if (item === "none") {
      if (queryParams.includes("&status=")) {
        setQueryParams(queryParams.replace(/&status=[^&]+/, ""));
      } else if (queryParams.includes("status=")) {
        setQueryParams(queryParams.replace(/status=.+?(&|$)/, ""));
      }
    } else {
      const NewParameter = `status=${item}`;
      if (queryParams.includes("&status=")) {
        let value = queryParams;
        value = value.replace(/&status=.[^&]+/, "");
        value = value === "" ? NewParameter : `${value}&${NewParameter}`;
        setQueryParams(value);
        //setQueryParams(queryParams.replace(/&status=.+?(&|$)/, `&${NewParameter}`))
      } else if (queryParams.includes("status=")) {
        let value = queryParams;
        value = value.replace(/status=.+?(&|$)/, "");
        value = value === "" ? NewParameter : `${value}&${NewParameter}`;
        setQueryParams(value);
      } else if (queryParams === "") {
        setQueryParams(NewParameter);
      } else {
        setQueryParams(`${queryParams}&${NewParameter}`);
      }
    }
    setShowStatusFilter(false);
  };

  const handleFromDate = (fromTS) => {
    const date = new Date(fromTS);
    const year = date.getFullYear();
    const month = date.getMonth() + 1;
    const day = date.getDate();
    const formattedDate = `${year}-${month}-${day}`;
    setFromTS(formattedDate);
    setDateError("");
  };
  const switchCalendars = ({ target }) => {
    const calendar = target.getAttribute("value");

    if (calendar === "from" && !showFromCalendar) {
      setShowFromCalendar(true);
      setShowToCalendar(false);
    }

    if (calendar === "to" && !showToCalendar) {
      setShowToCalendar(true);
      setShowFromCalendar(false);
    }
  };
  const closeCalendar = () => {
    if (showToCalendar) {
      //setToTS("none");
      setShowToCalendar(false);
      if (queryParams.includes("&end=")) {
        setQueryParams(queryParams.replace(/&end=.+?(&|$)/, ""));
      } else if (queryParams.includes("end=")) {
        setQueryParams(queryParams.replace(/end=.+?(&|$)/, ""));
      }
    }
    if (showFromCalendar) {
      //setFromTS("none");
      setShowFromCalendar(false);
      if (queryParams.includes("&start=")) {
        setQueryParams(queryParams.replace(/&start=.+?(&|$)/, ""));
      } else if (queryParams.includes("start=")) {
        setQueryParams(queryParams.replace(/start=.+?(&|$)/, ""));
      }
    }
  };
  const handleCalenderSubmission = () => {
    //add to link
    const toDate = new Date(toTS) 
    const fromDate = new Date(fromTS)
    if(toTS !== "none" && fromTS !== "none" && (toDate < fromDate )){
      setDateError("The 'end' date must be greater than the 'start' date");
      setFromTS("none");
      setToTS("none");
      closeCalendar();
      return;
    }
    if (toTS !== "none") {
      if (queryParams === "") {
        setQueryParams(`end=${toTS}`);
      } else {
        setQueryParams(`${queryParams}&end=${toTS}`);
      }
    }
    if (fromTS !== "none") {
      if (queryParams === "") {
        setQueryParams(`start=${fromTS}`);
      } else {
        setQueryParams(`${queryParams}&start=${fromTS}`);
      }
    }
    closeCalendar();
  };
  const handleToDate = (toTS) => {
    const date = new Date(toTS);
    const year = date.getFullYear();
    const month = date.getMonth() + 1;
    const day = date.getDate();
    const formattedDate = `${year}-${month}-${day}`;
    setToTS(formattedDate);
    setDateError("");
  };
  const handleOperationFilter = (item) => {
    setOperationField(item);
    if (item === "none") {
      if (queryParams.includes("&operation=")) {
        setQueryParams(queryParams.replace(/&operation=[^&]+/, ""));
      } else if (queryParams.includes("operation=")) {
        setQueryParams(queryParams.replace(/operation=.+?(&|$)/, ""));
      }
    } else {
      const NewParameter = `operation=${item}`;
      if (queryParams.includes("&operation=")) {
        let value = queryParams;
        value = value.replace(/&operation=[^&]+/, "");
        value = value === "" ? NewParameter : `${value}&${NewParameter}`;
        setQueryParams(value);
        //setQueryParams(queryParams.replace(/&operation=.+?(&|$)/, `&${NewParameter}`))
      } else if (queryParams.includes("operation=")) {
        let value = queryParams;
        value = value.replace(/operation=.+?(&|$)/, ``);
        value = value === "" ? NewParameter : `${value}&${NewParameter}`;
        setQueryParams(value);
      } else if (queryParams === "") {
        setQueryParams(NewParameter);
      } else {
        setQueryParams(`${queryParams}&${NewParameter}`);
      }
    }
    setShowOperation(false);
  };

  return (
    <section>
      {/* Logs */}
      <section className={styles.DetailsSection}>
        <div className="SectionTitle">Project logs</div>
        <div className={styles.Header}>
          <div className={styles.Heading}>Activity Feed</div>
          <div className={styles.SimpleForm}>
            <div className={styles.OuterFilterItem}>
            <div className={styles.DateSection}>
                  <div className={styles.DateItem}>
                      <div>Start:</div>
                      <DateInput
                        handleChange={handleFromDate}
                        showCalendar={showFromCalendar}
                        className={styles.dateField}
                        position={styles.CalenderFromposition}
                        dateValue={fromTS}
                        onClick={switchCalendars}
                        onCancel={closeCalendar}
                        onSubmit={handleCalenderSubmission}
                        value="from"
                      />
                    </div>
                    <div className={styles.DateItem}>
                      <div>End:</div>
                      <DateInput
                        handleChange={handleToDate}
                        showCalendar={showToCalendar}
                        position={styles.CalenderToposition}
                        className={styles.dateField}
                        dateValue={toTS}
                        onClick={switchCalendars}
                        onCancel={closeCalendar}
                        onSubmit={handleCalenderSubmission}
                        value="to"
                      />
                    </div>
                  </div>
                  <div className={styles.errorSection}>{dateError}</div>
            </div>
            <div className={styles.Filter}>
              <FilterIcon />
              <div className={styles.FilterText}>Filter</div>
              <DownArrow
                onClick={() => {
                  setFilterOpen(!filterOpen);
                }}
                className={styles.DropdownArrowSvg}
              />
            </div>
            {filterOpen && (
              <div className={styles.RelativeContainer}>
                <div className={styles.FilterDropdown}>
                  <div className={styles.FilterItemContainer}>
                    <div className={styles.FilterItem}>
                      Status
                      <div className={styles.SelectorBox}>
                        <div className={styles.SelectOption}>{statusField}</div>
                        <ArrowUpDDown
                          onClick={() => {
                            setShowStatusFilter(!showStatusFilter);
                          }}
                        />
                      </div>
                      {showStatusFilter && (
                        <div className={styles.InnerDropDown}>
                          <div
                            onClick={() => {
                              handleStatusFilter("none");
                            }}
                            className={styles.InnerDropDownItem}
                          >
                            none
                          </div>
                          {statusList.map((item, i) => (
                            <div
                              key={i}
                              className={styles.InnerDropDownItem}
                              onClick={() => {
                                handleStatusFilter(item);
                              }}
                            >
                              {item}
                            </div>
                          ))}
                        </div>
                      )}
                    </div>
                  </div>
                  <div className={styles.FilterItemContainer}>
                    <div className={styles.FilterItem}>
                      Operation
                      <div className={styles.SelectorBox}>
                        <div className={styles.SelectOption}>
                          {operationField}
                        </div>
                        <ArrowUpDDown
                          onClick={() => {
                            setShowOperation(!showOperation);
                          }}
                        />
                      </div>
                      {showOperation && (
                        <div className={styles.InnerDropDown}>
                          <div
                            onClick={() => {
                              handleOperationFilter("none");
                            }}
                            className={styles.InnerDropDownItem}
                          >
                            none
                          </div>
                          {operationsList.map((item, i) => (
                            <div
                              key={i}
                              className={styles.InnerDropDownItem}
                              onClick={() => {
                                handleOperationFilter(item);
                              }}
                            >
                              {item}
                            </div>
                          ))}
                        </div>
                      )}
                    </div>
                  </div>
                </div>
              </div>
            )}
          </div>
        </div>
        <div className={styles.CardArea}>
          {loading ? (
            <Spinner />
          ) : feedback !== "" ? (
            <div className={styles.NoResourcesMessage}>{feedback}</div>
          ) : (
            logs.map((item, index) => (
              <div key={index}>
                <div className={styles.TableRow}>
                  {item.status === "Success" ? (
                    <CheckMark className={styles.Success} />
                  ) : (
                    <Danger className={styles.Danger} />
                  )}
                  <div className={styles.Row}>
                    <div className={styles.RowCell}>
                      <User name="Demo" className={styles.UserAvatar} />
                      <div>
                        <div className={styles.ActivityEmail}>
                          {/* <b>{getUserEmail(item.user_id)}</b>: */}
                          <b>{item.user_id}</b>
                        </div>
                        <div className={styles.ActivityDate}>
                          {DisplayDateTime(new Date(item.creation_date))}
                        </div>
                      </div>
                      <div>
                        {item.operation}
                        <span className={styles.Entity}>
                          {item._id.$oid}
                        </span>{" "}
                        <span
                          className={
                            item.status === "Success"
                              ? styles.Success
                              : styles.Danger
                          }
                        >
                          {item.status}
                        </span>
                      </div>
                    </div>
                    <div className={styles.ActivityDescription}>
                      <div>{item.description}</div>
                    </div>
                  </div>
                </div>
                {index !== logs.length - 1 && <hr className={styles.hr} />}
              </div>
            ))
          )}
        </div>
      </section>
    </section>
  );
};

export default ActivityLogs;
