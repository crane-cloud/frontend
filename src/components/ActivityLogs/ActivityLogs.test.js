import React from 'react';
import { shallow } from 'enzyme';
import ActivityLogs from './';
import { handleGetRequest } from "../../apis/apis.js";

jest.mock("../../apis/apis.js");

describe('ActivityLogs', () => {
  test('renders the component', () => {
    const wrapper = shallow(<ActivityLogs projectID="123" />);
    expect(wrapper.exists()).toBe(true);
  });

  test('matches the snapshot', () => {
    const wrapper = shallow(<ActivityLogs projectID="123" />);
    expect(wrapper).toMatchSnapshot();
  });


});
