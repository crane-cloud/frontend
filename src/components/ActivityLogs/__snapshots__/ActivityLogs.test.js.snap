// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivityLogs matches the snapshot 1`] = `ShallowWrapper {}`;
